### PR TITLE
Feat/agent context instructions preset

### PR DESCRIPTION
**Pull request by TheovanKraay**
## Description

Part of #4200, and a follow-up to the discussion on #4259.

This is the delivery mechanism the maintainer proposed on #4259: instead of an extension manifest field that changes the agent's always-on behavior implicitly on install, always-on rules are delivered through an explicit, opt-in preset over `agent-context`. Core validates metadata only; the opt-in `agent-context` extension owns the writes; nothing touches the agent's context unless the user explicitly enables a preset.

- Presets (`src/specify_cli/presets/__init__.py`): `preset.yml` may declare `provides.instructions` (list of `{ file, description? }`), path-safe. An instructions-only preset (no templates) is valid. Metadata validation only.
- agent-context (all three twins): each installed and enabled preset's instruction block is composed into the managed section as a namespaced `<!-- SPECKIT PRESET:<id> START/END -->` sub-block. The Python twin is the single source of truth (`--emit-preset-blocks`); the bash and PowerShell twins delegate to it for byte-identical output. Path-unsafe, non-UTF-8, and marker-colliding payloads are skipped fail-closed. The PowerShell twin warns when no Python 3 with PyYAML is available and presets are installed, instead of silently omitting the blocks.
- `presets/example-always-on-rules/`: example instructions-only preset.
- Docs: `docs/reference/presets.md`, `presets/PUBLISHING.md`.

Ownership and consent: the preset is the standalone unit (`specify preset add` / `disable` / `remove`), so installing an extension does not by itself change the agent. Enabling the preset is the explicit opt-in. An extension author can also distribute the preset together with the extension and `agent-context` as a bundle, so `specify bundle install` sets everything up in one previewable, consented step; that uses the existing bundle mechanism and lives with the extension, so it is out of scope here.

## Testing

- [x] Tested locally with `uv run specify --help`
- [x] Ran existing tests with `uv sync && uv run pytest`
- [x] Tested with a sample project (if applicable)

New `tests/extensions/test_preset_instructions.py` (16) and preset-parity tests in `tests/extensions/test_update_agent_context_python_parity.py` (Python-vs-Bash on CI, Python-vs-PowerShell locally, non-ASCII payload). No regressions across the preset and extension suites (771 passed, 144 skipped). Verified end to end via both the Python and PowerShell twins: `specify init` -> `specify preset add --dev presets/example-always-on-rules` -> `specify extension add --dev extensions/agent-context` -> agent-context update composes the block into `.github/copilot-instructions.md`.

## AI Disclosure

- [ ] I **did not** use AI assistance for this contribution
- [x] I **did** use AI assistance (describe below)

Implemented with GitHub Copilot (agentic): preset validation, the agent-context composition and its bash/powershell twins, the example preset, the tests, and the docs were written with AI assistance and reviewed by me.

### CODE REVIEW

File: extensions/agent-context/scripts/python/update_agent_context.py [hunk 1]
```diff
@@ -201,18 +207,128 @@ def _resolved_rel(p: Path) -> Path | None:
...
+        try:
+            with open(manifest, "r", encoding="utf-8") as fh:
+                pdata = yaml.safe_load(fh)
+        except Exception:
+            continue
+        provides = pdata.get("provides") if isinstance(pdata, dict) else None
+        instructions = provides.get("instructions") if isinstance(provides, dict) else None
+        if not isinstance(instructions, list):
+            continue
+        preset_root = (presets_dir / preset_id).resolve()
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply extensions/agent-context/scripts/python/update_agent_context.py]: Good catch. The registry is untrusted input, so joining the preset id straight onto `.specify/presets` let an absolute key or a `..`/separator id (or a symlinked directory) point the manifest read outside the presets root.

Fixed in f660e11: the collector now rejects any id that is not a simple name (must match `^[a-z0-9][a-z0-9._-]*$`, so no separators, `..`, or absolute/drive forms) and, after resolving, confirms the preset directory still lives inside the resolved presets root via `relative_to` before it opens `preset.yml`. A crafted key or symlink is skipped instead of read.

Added `test_unsafe_registry_preset_id_skipped`, which injects `../../evil` and `/abs-evil` into the registry and asserts the good preset still composes while nothing from the escaped paths appears in the section.

File: src/specify_cli/presets/__init__.py [hunk 2]
```diff
@@ -499,6 +501,45 @@ def _validate(self):
...
+        # Validate instructions (optional): each entry declares a markdown rule
+        # block that the opt-in agent-context extension composes into the
+        # always-on context file when this preset is enabled. Path-safety mirrors
+        # the template file checks above.
+        if has_instructions:
+            instructions = provides["instructions"]
+            if not isinstance(instructions, list):
+                raise PresetValidationError(
+                    "Invalid provides.instructions: expected a list"
+                )
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply src/specify_cli/presets/__init__.py:514]: Agreed. An empty `provides.instructions` list passed validation but contributed nothing, which is inconsistent with how we already reject empty templates.

Fixed in f660e11: validation now raises `provides.instructions must not be empty` when the list is present but empty, mirroring the empty-templates rejection. Added `test_empty_instructions_list_rejected`.

File: src/specify_cli/presets/__init__.py [hunk 3]
```diff
@@ -499,6 +501,45 @@ def _validate(self):
...
+                if (
+                    file_path.startswith("/")
+                    or "\\" in file_path
+                    or os.path.isabs(normalized)
+                    or normalized.startswith("..")
+                ):
+                    raise PresetValidationError(
+                        f"Invalid instruction file path '{file_path}': "
+                        "must be a relative path within the preset directory"
+                    )
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply src/specify_cli/presets/__init__.py]: Right, the ad-hoc check let non-portable forms through (empty, whitespace, `.`, directory-only, and Windows drive-relative like `C:rules.md`).

Fixed in f660e11: instruction file paths now go through the shared `relative_extension_path_violation` policy in `_utils.py`, the same one the extension path validation uses, so all of those forms are rejected at metadata-validation time with a specific reason. Added a parametrized `test_instruction_path_non_portable_rejected` covering empty, whitespace, surrounding whitespace, `.`, directory-only, drive-relative, and backslash forms.

File: extensions/agent-context/scripts/python/update_agent_context.py [hunk 4]
```diff
@@ -201,18 +207,140 @@ def _resolved_rel(p: Path) -> Path | None:
...
+                continue
+            target = (preset_root / rel).resolve()
+            try:
+                target.relative_to(preset_root)
+            except ValueError:
+                continue
+            if not target.is_file():
+                continue
+            try:
+                text = target.read_text(encoding="utf-8").strip()
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply extensions/agent-context/scripts/python/update_agent_context.py:345]: Good call. The composed managed section is re-sent as agent context on every request, so an oversized instruction file (a bundled archive member or an unbounded `--dev` source) could bloat it without bound.

Fixed in a899821: added a deliberately small budget in the collector. Any single file over a per-file cap (32 KiB) is skipped with a warning, and the on-disk size is checked with `stat()` before the file is read so a huge member is never allocated into memory. A running aggregate cap (64 KiB across all presets) stops composition once reached, and each skip logs which preset and why.

Boundary coverage added: `test_instruction_file_at_limit_included` (a file exactly at the per-file cap is kept, since the check is strictly greater), `test_oversized_instruction_file_skipped` (over-cap file skipped, other presets still compose), and `test_aggregate_instruction_budget_enforced` (three under-cap presets where the third crosses the aggregate cap and is dropped in id order). The bash/ps1 twins delegate to this collector via `--emit-preset-blocks`, so the budget applies uniformly.

File: tests/extensions/test_preset_instructions.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+
+def test_unsafe_registry_preset_id_skipped(tmp_path):
+    # The registry is on disk and untrusted: an id with separators/traversal or
+    # an absolute/drive form must be skipped, not resolved and read.
+    _configure_agent_context(tmp_path)
+    _install_preset(tmp_path, "good", RULES_A)
+    reg_path = tmp_path / ".specify" / "presets" / ".registry"
+    reg = json.loads(reg_path.read_text(encoding="utf-8"))
+    reg["presets"]["../../evil"] = {"version": "1.0.0", "enabled": True}
+    reg["presets"]["/abs-evil"] = {"version": "1.0.0", "enabled": True}
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply tests/extensions/test_preset_instructions.py:337]: Right, the previous version only injected registry keys, so collection stopped at `manifest.is_file()` and the test would have passed even without the guards.

Fixed in a899821: the test now materializes a real out-of-root preset (a `preset.yml` plus an instruction payload tagged `PWNED_TRAVERSAL_PAYLOAD`) at exactly the location the resolved `../../evil` key points at, and asserts that payload is not composed. So if the id-format or containment guard is removed, collection resolves and reads it and the test fails.

I also added `test_symlinked_preset_dir_escaping_root_skipped`: a valid simple id (`linked`) whose directory is a symlink pointing outside `.specify/presets`. That case passes the id regex, so it specifically exercises the resolved-containment check; it asserts the out-of-root payload (`PWNED_SYMLINK_PAYLOAD`) is not composed, and skips only where the platform disallows symlink creation.

**Comment by mnriem** [comment]
Note that you should be able to define this completely as a preset and you should not need to adjust the extension scripts. The preset can wrap/replace those scripts (what you pick depends on your need).

File: extensions/agent-context/scripts/bash/update-agent-context.sh [hunk 6]
```diff
@@ -354,6 +355,13 @@ trap 'rm -f "$TMP_SECTION"' EXIT
   if [[ -n "$PLAN_PATH" ]]; then
     echo "at $PLAN_PATH"
   fi
+  # Always-on instruction blocks contributed by enabled presets (#4200).
+  # Delegated to the python twin's --emit-preset-blocks so all three twins emit
+  # byte-identical block text from a single implementation.
+  _PRESET_BLOCKS="$("$_python" "$_SCRIPT_DIR/../python/update_agent_context.py" --emit-preset-blocks --marker-start "$MARKER_START" --marker-end "$MARKER_END" 2>/dev/null || true)"
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply extensions/agent-context/scripts/bash/update-agent-context.sh]: Good catch, and this was a real destructive-failure bug. The `2>/dev/null || true` around the emit call both hid the composer's warnings and turned any nonzero exit into an empty result, which then rewrote the managed section with previously composed preset blocks silently dropped.

Fixed in 7107d8d: the call now captures only stdout (so the composer's oversized/marker-colliding/skipped warnings reach stderr) and checks the exit code, aborting the update on a nonzero exit so the section is never rewritten on failure. I moved it out of the section-building block so the abort happens before the temp section is assembled.

Verified locally: the happy path composes the block; an oversized entry warns on stderr and exits 0 so the update proceeds without it (a skip is not a failure); a hard composer failure aborts before any rewrite.

File: extensions/agent-context/scripts/powershell/update-agent-context.ps1 [hunk 7]
```diff
@@ -457,6 +457,52 @@ $lines = @($MarkerStart,
...
+        [Console]::OutputEncoding = [System.Text.Encoding]::UTF8
+        $emitted = (& $pyForBlocks $pyTwin --emit-preset-blocks --marker-start $MarkerStart --marker-end $MarkerEnd 2>$null | Out-String)
+    } finally {
+        [Console]::OutputEncoding = $prevOutEnc
+    }
+    if ($emitted) {
+        $emitted = ($emitted -replace "`r`n", "`n") -replace "`r", "`n"
+        $emitted = $emitted.TrimEnd("`n")
+        foreach ($bl in ($emitted -split "`n")) { $lines += $bl }
+    }
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply extensions/agent-context/scripts/powershell/update-agent-context.ps1:525]: Agreed, same destructive failure mode as the bash twin. `2>$null` discarded the composer's warnings and `$LASTEXITCODE` was never checked, so an emitter error was treated as "no blocks" and the upsert dropped existing preset instructions.

Fixed in 7107d8d: stderr is now routed to a temp file, surfaced to the console after the call (so warnings are preserved separately from the captured stdout), and the exit code is checked. A nonzero exit writes an error and aborts before `$Section` is assembled, so the managed section is never rewritten on failure. The UTF-8 OutputEncoding wrapper for stdout capture is kept.

Verified locally on Windows via a direct run of the PowerShell twin: the block composes on the happy path, and the emit path returns 0 with a stderr warning when an entry is skipped.

File: presets/PUBLISHING.md [hunk 8]
```diff
@@ -76,6 +76,9 @@ provides:
       file: "templates/spec-template.md"
       description: "Custom spec template"
       replaces: "spec-template"
+  instructions:                     # Optional: always-on rule blocks composed
+    - file: "instructions/best-practices.md"   # by the opt-in agent-context extension
+      description: "Always-on engineering rules"
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply presets/PUBLISHING.md:81]: Right, since core validates metadata only, a preset can install cleanly yet contribute nothing at runtime, and that was not documented.

Fixed in 7107d8d: `presets/PUBLISHING.md` now documents, right after the manifest example, the runtime constraints the agent-context composer applies to each `provides.instructions` file: it must exist inside the preset directory and be valid UTF-8, must not contain a managed-section marker, must be at or below 32 KiB per file, and must fit the 64 KiB aggregate budget across all enabled presets. It notes that entries failing any of these are dropped fail-closed and logged to stderr while the rest still compose, and explains the budget exists because the managed section is re-sent as agent context on every request.

File: extensions/agent-context/scripts/python/update_agent_context.py [hunk 9]
```diff
@@ -201,18 +216,164 @@ def _resolved_rel(p: Path) -> Path | None:
...
+                )
+                continue
+            if marker_start in text or marker_end in text or _SPECKIT_MARKER_RE.search(text):
+                _err(
+                    f"agent-context: skipping instructions from preset '{preset_id}': "
+                    "content contains a managed section marker."
+                )
+                continue
+            total_bytes += entry_bytes
+            parts.append(text)
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply extensions/agent-context/scripts/python/update_agent_context.py]: Good catch, the counter was undercounting. It summed only each stripped payload, so a preset that references a tiny file thousands of times rendered well past 64 KiB while the counter stayed low (your 22,000-references example).

Fixed in 764c705: the accounting now adds the bytes each entry actually contributes to the rendered section, not just its payload. The first entry of a preset adds the surrounding marker block (`<!-- SPECKIT PRESET:<id> START/END -->` plus its newlines), and every later entry adds the two-byte blank-line separator. So the cap now bounds the composed always-on context, not an under-estimate of it.

Added `test_aggregate_budget_counts_wrapper_and_separators`: a single preset referencing a 1-byte file 25,000 times (raw sum under the cap) and asserts the composed section stays within `_MAX_INSTRUCTION_TOTAL_BYTES` and is truncated below the full entry count, so it fails if the overhead is dropped from the accounting.

File: presets/PUBLISHING.md [hunk 10]
```diff
@@ -76,12 +76,23 @@ provides:
...
   - "workflow"
 ```
 
+**About `provides.instructions`** (optional): each entry names a Markdown file whose contents the opt-in `agent-context` extension composes into the always-on managed section of the agent context file (for example `.github/copilot-instructions.md`) on the next `agent-context` update. Core validates only the metadata (the entry shape and a portable relative `file` path), so an accepted preset can still contribute nothing at runtime unless every referenced file also satisfies the composer's constraints:
+
+- The file must exist inside the preset directory and be valid UTF-8.
+- It must not contain a managed-section marker (`<!-- SPECKIT ... -->`); such payloads are skipped to avoid corrupting the section.
+- Each file must be at or below 32 KiB, and the combined instructions across all enabled presets must fit a 64 KiB aggregate budget; the managed section is re-sent as agent context on every request, so the budget is deliberately small.
+
+Entries that fail any of these are dropped (fail-closed) and logged to stderr; the remaining ones still compose.
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply presets/PUBLISHING.md]: Right, the doc over-promised. The collector warned on oversized/over-budget/marker-collision but silently skipped missing, unreadable, non-UTF-8, and path-unsafe files.

Fixed in 764c705 by doing both things you suggested. The collector now emits an stderr warning for a missing or unreadable/non-UTF-8 instruction file (these are author errors worth surfacing). Path-unsafe entries (absolute, parent-traversal, or a target that escapes the preset directory) stay silent on purpose, since that is a security fail-closed decision and I did not want to give traversal attempts useful feedback. The doc now states exactly that split: which skips warn versus which are intentionally silent, and also notes the aggregate budget counts the rendered wrappers/separators.

File: extensions/agent-context/scripts/powershell/update-agent-context.ps1 [hunk 11]
```diff
@@ -457,6 +457,65 @@ $lines = @($MarkerStart,
...
+        try {
+            $preg = Get-Content -LiteralPath $presetReg -Raw -Encoding UTF8 | ConvertFrom-Json
+            $enabled = @($preg.presets.PSObject.Properties | Where-Object { $_.Value.enabled -ne $false })
+            if ($enabled.Count -gt 0) {
+                [Console]::Error.WriteLine("agent-context: Python 3 with PyYAML not found; preset always-on instruction blocks (provides.instructions) were NOT composed. Base context section written.")
+            }
+        } catch { }
+    }
+}
+if ($pyForBlocks -and (Test-Path -LiteralPath $pyTwin)) {
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply extensions/agent-context/scripts/powershell/update-agent-context.ps1]: Good edge case. If Python was found but the sibling composer file was gone (a partial or corrupt install), the combined guard was just false and the script proceeded to rewrite the managed section without any preset blocks, silently erasing existing instructions.

Fixed in 719d319: when Python is available the ps1 twin now treats a missing composer as a hard failure, writing an error and aborting before the section is rewritten, exactly like the nonzero-exit path. The bash twin already covers this case, since a missing composer file makes python exit nonzero and the exit-code check aborts. Verified the ps1 parses clean and the happy path still composes.
